File: Python/BellmanFord.py
# robin-0 ( Nabil Ibtehaz)

'''
	This code implements the Bellman-Ford Algorithm
	to solve the single source shortest path problem.

'''

class Edge:

	''' 
		This is the class to model graph edges
		it has 3 attributes startingNode , 
		endingNode and weight
	'''

	# constructor
	def __init__(self,startingNode,endingNode,weight):
		self.startingNode=startingNode
		self.endingNode=endingNode
		self.weight=weight


def BellmanFord(source):

	'''
		This function implements the Bellman Ford algorithm
		to find the single source shortest path problem

		It takes the source vertex as input and returns a
		list containing the lengths of shortest paths to 
		other vertices
	'''

	shortestPaths = [ None ] * nVertices

	shortestPaths[source]=0   # path from source to source is 0 :p

	for i in range(nVertices-1):
		for j in edgeList:

			if( shortestPaths [ j.startingNode ] !=None ):

				if(shortestPaths [j.endingNode] ==None):

					shortestPaths [j.endingNode] = shortestPaths [ j.startingNode ] + j.weight

				else :
					# Relaxation
					shortestPaths [j.endingNode] = min( shortestPaths [j.endingNode] , shortestPaths [ j.startingNode ] + j.weight	)				

	# check for negative cycle

	for j in edgeList:

		if( shortestPaths [ j.startingNode ] !=None and shortestPaths [ j.endingNode ] !=None ):
				
			if ( shortestPaths [j.endingNode] > ( shortestPaths [ j.startingNode ] + j.weight ) ):

				print("Not Possible Negative Cycle Exists")

				return None

	return shortestPaths



if __name__ == "__main__":

	'''
		This is the main function.

		A simple UI is provided
	'''

	print('Vertices are 0-indexed')
	nVertices = int(input('Number of Vertices: '))

	edgeList = []

	nEdges = int(input('Number of Edges: '))
	
	for i in range(nEdges):


		print('Enter Edge ' +str(i+1) )

		startingNode = int(input('Starting Node: '))
		endingNode = int(input('Ending Node: '))
		weight = int(input('Weight: '))

		edgeList.append(Edge(startingNode,endingNode,weight))


	print("****************************************\n")
	print("Please select one of the following\n")
	choice = int(input('1. Find Shortest Path From Sources\n2. End\n'))

	while(choice==1):
		sourceNode=int(input('Source Node(0-indexed) : '))
		print(BellmanFord(sourceNode))

		print("****************************************\n")
		print("Please select one of the following\n")
		choice = int(input('1. Find Shortest Path From Sources\n2. End\n'))

File: CPP/BellmanFord.cpp
#include <algorithm>
#include <iostream>

#define INF 100000000

using namespace std;

typedef struct edge {
	int u, v, w;
} edge;

vector<edge> edges;
vector<int> dis;
bool has_negative_cycle;
int n_nodes;

void bellmanford(int src) {
	dis[src] = 0;
	has_negative_cycle = false;
	edge e;
	for(int i = 1;i < n_nodes;i++) {
		for(vector<edge>::iterator it = edges.begin();it != edges.end();it++) {
			e = *it;
			if(dis[e.u] != INF) {
				dis[e.v] = min(dis[e.u] + e.w, dis[e.v]);
			}
		}
	}
	// detect negative cycle
	for(vector<edge>::iterator it = edges.begin();it != edges.end() && !has_negative_cycle;it++) {
		e = *it;
		if(dis[e.u] != INF && dis[e.u] + e.w < dis[e.v]) {
			has_negative_cycle = true;
		}
	}
}

int main(int argc, char const *argv[]) {
	int n_edges, u, v, w;
	cin >> n_nodes;
	dis.assign(n_nodes, INF);
	cin >> n_edges;
	while(n_edges--) {
		cin >> u >> v >> w;
		edges.push_back({u, v, w});
	}
	// input source node
	cin >> u;
	bellmanford(u);
	if(has_negative_cycle) {
		cout << "Negative cycle detected" << endl;
	}
	else {
		cout << "No negative cycle detected" << endl;
		for(int i = 0;i < n_nodes;i++) {
			cout << i << ": " << dis[i] << endl;
		}
	}

	return 0;
}
